File: year_2023/day_06/day06.py
import re

from more_itertools import transpose

from tools import loader, parsers


def race(data: list[str], part2: bool = False) -> int:
    """
    >>> print(race(parsers.lines('test.txt'), part2=False))
    288
    >>> print(race(parsers.lines('test.txt'), part2=True))
    71503"""
    records = list(transpose(list(map(int, re.findall(r'\d+', i.split(':')[1]))) for i in data))
    if part2:
        records = [tuple(int(re.sub(r'\s+', '', i.split(':')[1])) for i in data)]
    wins = 1
    for time, record in records:
        wins *= sum(record < (time - hold) * hold for hold in range(1, time + 1))
    return wins


print(race(parsers.lines(loader.get()), part2=False))  # 138915
print(race(parsers.lines(loader.get()), part2=True))  # 27340847

File: year_2023/day_06/day06.go
package main

import (
	"github.com/Nyaaa/advent-of-code/tools"
	"fmt"
	"strconv"
	"strings"
)

func solve(input string, part2 bool) int {
	data := tools.ReadLines(input)
	times := strings.Fields(data[0])[1:]
	distances := strings.Fields(data[1])[1:]
	wins := 1

	if part2 {
		times = []string{strings.Join(times, "")}
		distances = []string{strings.Join(distances, "")}
	}

	for i := 0; i < len(times); i++ {
		time, _ := strconv.Atoi(times[i])
		distance, _ := strconv.Atoi(distances[i])
		w := 0

		for i := 1; i <= time+1; i++ {
			if distance < (time-i)*i {
				w++
			}
		}

		wins *= w
	}

	return wins
}

func main() {
	data := tools.GetData(2023, 6)

	fmt.Println(solve("test.txt", false), solve("test.txt", true))
	fmt.Println(solve(data, false), solve(data, true))
}
